sycl/test-e2e/Basic/dependent_event_async_exception.cpp

```diff
@@ -0,0 +1,109 @@
+// RUN: %{build} -o %t.out
+// RUN: %{run} %t.out
+
+//==----- dependent_event_async_exception.cpp - Test for event async exceptions
+//-----==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include <sycl/sycl.hpp>
+#include <unordered_set>
+
+struct test_exception {
+  std::string name;
+};
+
+class test_exception_handler {
+public:
+  test_exception_handler()
+      : queue{[this](sycl::exception_list el) { capture(std::move(el)); }} {}
+  sycl::queue &get_queue() { return queue; }
+
+  bool has(const std::string &name) const {
+    return captured_exceptions.count(name) != 0;
+  }
+
+  size_t count() const { return captured_exceptions.size(); }
+
+  void clear() { captured_exceptions.clear(); }
+
+private:
+  std::unordered_set<std::string> captured_exceptions;
+  sycl::queue queue;
+
+  void capture(sycl::exception_list el) {
+    for (auto &e : el) {
+      try {
+        std::rethrow_exception(e);
+      } catch (test_exception &te) {
+        captured_exceptions.insert(te.name);
+      }
+    }
+  }
+};
+
+static sycl::event
+make_throwing_host_event(sycl::queue &queue, std::string name,
+                         const std::vector<sycl::event> &dependencies = {}) {
+  return queue.submit([name, &dependencies](sycl::handler &cgh) {
+    for (auto &dep : dependencies) {
+      cgh.depends_on(dep);
+    }
+    cgh.host_task([name](auto) { throw test_exception{name}; });
+  });
+}
+
+int main() {
+  {
+    test_exception_handler teh1;
```

Comment:
I think it could be worth also having a test for the case that @gmlueck highlighted in the [spec PR](https://github.com/KhronosGroup/SYCL-Docs/pull/520#discussion_r1440744161 ) where there is no dependency chain between the two kernels but they are enqueued to the same queue.

It was decided that in this case we should all errors caught by the `queue`, so if there is a call to `wait` on the `queue` then both exceptions should be thrown.